Frameworks/CoreText/DWriteWrapper_CTFont.mm

```diff
@@ -208,6 +208,173 @@ HRESULT _DWriteCreateFontFaceWithFontDescriptor(CTFontDescriptorRef fontDescript
+                                                  forceMatch);
+    } else if (CFStringCompare(key, kCTFontStyleNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByStyleName(fonts, static_cast<CFStringRef>(value), forceMatch);
+    } else if (CFStringCompare(key, kCTFontTraitsAttribute, 0) == kCFCompareEqualTo) {
+        __FilterFontsByTraits(fonts, static_cast<CFDictionaryRef>(value), forceMatch);
+        return S_OK;
+    } else {
+        // Other attributes are currently unsupported by CTFont
+        UNIMPLEMENTED();
+        return S_FALSE;
+    }
+}
+
+#pragma endregion // Helper Filter Methods
+
+/**
+ * Creates a CFArray of CTFontDescriptors values for all fonts that match the given attributes best and completely match mandatoryKeys
+ */
+HRESULT _DWriteCreateMatchingFontDescriptors(CFDictionaryRef attributes, CFSetRef mandatoryKeys, CFArrayRef* matchingNames) {
+    *matchingNames = nullptr;
+
+    // Get list of all possible fonts
+    std::vector<ComPtr<IDWriteFont>> fonts;
+    RETURN_IF_FAILED(_DWriteGetAllFonts(fonts));
+
+    if (mandatoryKeys) {
+        // Filter by the mandatory keys first
+        CFIndex count = CFSetGetCount(mandatoryKeys);
+        std::vector<CFStringRef> keys(count);
+        CFSetGetValues(mandatoryKeys, (const void**)keys.data());
+        for (auto& key : keys) {
+            if (!CFDictionaryContainsKey(attributes, key)) {
+                // Mandatory key does not exist, so no fonts can match
+                return S_OK;
+            }
+
+            auto value = woc::MakeAutoCF<CFTypeRef>(CFDictionaryGetValue(attributes, key));
+            RETURN_IF_FAILED(__FilterFontsByKeyAndValue(fonts, key, value, true));
+            if (fonts.empty()) {
+                // No fonts that match mandatory keys, can stop filtering now
+                return S_OK;
+            }
+        }
+    }
+
+    // Left with list of fonts matching mandatoryKeys, can now try to match any non-mandatory attributes
+    CFIndex count = CFDictionaryGetCount(attributes);
+    if (count > 0L) {
+        std::vector<CFStringRef> keys(count);
+        std::vector<CFTypeRef> values(count);
+        CFDictionaryGetKeysAndValues(attributes, (const void**)keys.data(), (const void**)values.data());
+        for (size_t i = 0; i < count; ++i) {
+            if (!mandatoryKeys || !CFSetContainsValue(mandatoryKeys, keys[i])) {
+                RETURN_IF_FAILED(__FilterFontsByKeyAndValue(fonts, keys[i], values[i], false));
+            }
+        }
+    }
+
+    std::unique_ptr<CFTypeRef[]> outDescriptors(new CFTypeRef[fonts.size()]);
+    for (size_t i = 0; i < fonts.size(); ++i) {
+        BOOL exists;
+        ComPtr<IDWriteLocalizedStrings> dwriteString;
+        RETURN_IF_FAILED(fonts[i]->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME, &dwriteString, &exists));
+        CFStringRef name = _CFStringFromLocalizedString(dwriteString.Get());
```

Comment:
fwiw I read it as an autorelease value. if I write it like [NSString stringFromLocalizedString:...] that looks a lot like an autorelease even though it doesn't have 'get', right?

---
In reply to: [104537218](https://github.com/Microsoft/WinObjC/pull/2171#discussion_r104537218) [](ancestors = 104537218)